### PR TITLE
stdenv: single make jobserver across multiple nix builds

### PR DESCRIPTION
**Pull request by pennae**
###### Motivation for this change

`make -jN -lN` in stdenv is a very blunt instrument. it works well when max-jobs=1, but as nix-level paralellism increases it becomes increasingly deficient. starting from a low-load situation we start max-jobs * N compilers, loadavg goes through the roof, the `-lN` load limit kicks in and inhibits new compilers starting until loadavg has fallen below N—at which point all make instances spawn a lot of new compilers and loadavg goes through the roof again. this oscillation leaves the system underutilized in low phases and overcommitted in high phases.

testing the current stdenv against a jobserver with 26 tokens on a 12C/24T machine shows that parallel builds of llvm_{8..11} run about 7% faster (35:52min for stdenv, 33:30min with jobserver), a larger build of llvm{5..13} is about about 11% faster (1:27h for stdenv, 1:17h with jobserver). (<del>removing the `-l` from stdenv also improves utilization but is less efficient. preliminary testing here shows that `-l${1.5 * N}` may be a good alternative to `-lN` as used currently, #141266 could be a good vector to go for that instead of this whole mess.</del> [more testing says that `-l2N` would be a minimum to get better utilization, but so far every `-l` setting we've tried has produced *some* underutilization except excessive large numbers like 6N or higher])

nothing in here should be regarded as a final suggestion in any way, it's more of a "hey look, this might just work". as such it's extremely rough around the edges, eg to use the jobserver the experimenter currently has to bring a `/jobserver` fifo filled with tokens into the nix sandbox:
```sh
nix-build -E 'with import ./. {}; pkgs.callPackage ./pkgs/os-specific/linux/nixos-jobserver {}'
touch /tmp/jobserver
sudo ./result -t 24 -g nixbld /tmp/jobserver&
nix build --extra-sandbox-paths "/jobserver=/tmp/jobserver" #...
```

is this something worth pursuing? a 10% speedup for hydra does seem tempting

todos before this is more generally usable:
 - [x] re-add $NIX_BUILD_CORES support (ioctl on the jobserver fd should do it)
 - [x] maybe port from cuse to fuse to not require root for everything (mmap problems may be solved by reporting jobserver file size as 0 at all times, needs testing)
 - [x] usability of tools (error messages, runtime configuration, etc)
 - [x] add support for other build systems (cargo can honor MAKEFLAGS, maybe others too)
 - [x] nixos module
 - [x] make configurable which env var(s) jscall adds jobserver information to
 - [ ] metrics?
 - [ ] \<your suggestion here\>

cc @vcunat @Artturin

### CODE REVIEW

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -268,6 +268,16 @@ else let
...
       inherit outputs;
+    } // lib.optionalAttrs (stdenv.hasCC) {
+      jscall = derivation {
+        inherit (stdenv.hostPlatform) system;
+        name = "jscall";
+        src = ../generic/jscall.c;
+        builder = stdenv.shell;
+        args = [ "-e" "-c" ''
+            ${stdenv.cc}/bin/cc -o $out $src
+        '' ];
```
Comment (SuperSandro2000) [inline pkgs/stdenv/generic/make-derivation.nix:279]:
```suggestion
        args = [
          "-e"
          "-c"
          "${stdenv.cc}/bin/cc -o $out $src"
        ];
```

**Comment by pennae** [comment]
the token loss problem could be solved on linux with a little fuse filesystem that emulates pipes, tracks how many tokens were taken from each file description of a pipe, and returns missing tokens when a pipe file description is released. darwin also has a fuse implementation, so there may just be a chance that if this works on linux it could also work reasonably well on darwin at some point.

**Comment by pennae** [comment]
updated with a CUSE-based jobserver fifo-alike. this variant survives nix-daemon killing builds hard without losing tokens, and in theory it could be implemented as a FUSE filesystem too (eg for darwin, or if the CUSE variant is deemed too exotic). at the moment it's CUSE mostly because FUSE doesn't seem to have a way to forbid mmapping of files, and while mmapping a token fd wouldn't *break* anything it *would* consume 4k tokens in one fell swoop and not give them back for a good while.

**Comment by pennae** [comment]
updated with support for cargo, reintroduced per-build core limits, and added a nixos module. getting there :crossed_fingers:

File: pkgs/development/compilers/edk2/default.nix [hunk 2]
```diff
@@ -25,7 +23,7 @@ else
   throw "Unsupported architecture";
 
 buildStdenv = if stdenv.isDarwin then
-  overrideCC clangStdenv [ clang_9 llvmPackages_9.llvm llvmPackages_9.lld ]
```
Comment (expipiplus1) [inline pkgs/development/compilers/edk2/default.nix:28]: Did this slip in by mistake?

**Comment by expipiplus1** [comment]
This is perhaps outside the initial scope, but wouldn't one want to have the nix daemon itself get slots from the jobserver? (As the jobserver protocol says workers have one implicit job just by virtue of being started, and obviously non-jobserver-compatible builds take a core)

File: pkgs/development/compilers/edk2/default.nix (hunk 2 above)
Comment (pennae) [reply pkgs/development/compilers/edk2/default.nix:28]: nope. without this change this package will fail to build trying to build stdenv.jscall (because after this override stdenv.cc is a list, which is illegal. the jscall drv interpolates stdenv.cc into a bash string directly, thus failing here)

**Comment by pennae** [comment]
yes, in an ideal world we'd have the jobserver as part of nix-daemon and have nix-daemon itself pull tokens for the jobs it starts. that's a whole lot of work we're not willing to do yet though while it's not proven that there's interest in this feature :D

**Comment by L-as** [comment]
I am not sure I am for this patch. This has the potential to introduce many subtle reproducibility issues, not to mention it's not very pure. Since we're so close, an approach that utilises Nix as explained in RFC 92 is much better IMO.

**Comment by pennae** [comment]
it is impure currently, but i think all user-visible impurity could be resolved by putting a jobserver into nix-daemon directly. reproducibility also shouldn't a problem anymore then, since the only observable difference between builds with vs without an external jobserver would be the result of stat() on jobserver fds.

also we can't quite see how implementing rfc 92 would solve oversubscription. turning make or ninja buildsystems into a lot of computed derivations, one per input file, sounds fine and would be preferrable to this. but what about linkers that can use a jobserver to speed up linking, or build systems like cargo that parallelize internally and externally act on the levels of entire projects?

we'd be delighted about solutions to the problems that don't require hacks like this, but for time being it does seem to us that this is the easiest way to solve oversubscription. we suspect that this is not even be a problem most people have, it'd probably be more useful to speed up hydra (maybe) or for developers building lots of things (where reproducibility is less immediately important, and better efficiency for builds goes a long way). we just don't know, that's why it's a draft :slightly_smiling_face:

**Comment by L-as** [comment]
When you have smaller derivations, one thread per derivation is enough. Build systems like cargo would have their build logic replaced by derivations.

**Comment by pennae** [comment]
if that is possible, all for it. if it's not possible soon though i would at least propose this as an optional stopgap for those who are heavily affected by oversubscription, like we've seen with cargo and ghc.

**Comment by expipiplus1** [comment]
I opened a patch for GHC adding jobserver support: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7000. Hopefully I can get some encouraging performance numbers with nix soon.

> I am not sure I am for this patch. This has the potential to introduce many subtle reproducibility issues, not to mention it's not very pure. Since we're so close, an approach that utilises Nix as explained in RFC 92 is much better IMO.

We allow NIX_BUILD_CORES in, and that probably has similar scope for reproducibility issues, right?

@pennae's patch is very much opt in, if the user isn't running a job-server then it's do-nothing. It also has to be enabled explicitly for any tool, which can hopefully be verified to not behave improperly.

**Comment by Profpatsch** [comment]
> a 10% speedup for hydra does seem tempting

A 10% speedup for making things impure and fairly uncontrollable? Not worth it if you ask me.

Bazel has had support for jobservers for a while, but integrating takes human-*years* to get right and bug-free. The complexity is devillish.

**Comment by Profpatsch** [comment]
Instead of throwing hundreds of hours at the problem, how about we spend the equivalent in money to increase the hydra build farm by 10% instead

**Comment by L-as** [comment]
Other things to speed up building:
- Improve the speed of libstore itself
- Split up derivations more (with dynamic derivations and recursive nix possibly)
- Use faster compilers, e.g. tinycc, for non-performance-sensitive programs.
- Use mold as a linker.
- Use clang elsewhere.

**Comment by pennae** [comment]
the problem we're trying to solve here is that multiple instances of build systems that use system average load as a metric to limit how many jobs they run creates fairly hefty oscillations of under- and overutilization of the builder. dynamic derivations and recursive nix have a good chance of being able to solve that, once every major build system has a translator to a derivation per build step. parallel linkers may even be detrimental in such cases unless we teach nix something like job weight, to avoid the rust situation nowadays where we start `ncpu` cargo instances and each of those starts `ncpu` build threads.

we're not disagreeing that all of those things are better solutions than this here. they're just *far away*.

**Comment by L-as** [comment]
They are probably much closer than you think. If we fix https://github.com/L-as/ninja2nix (see issues), we could begin using dynamic derivations soon after.

**Comment by pennae** [comment]
ninja would greatly benefit from your tooling but not from this PR, so all for that! :+1: but that's only ninja, and we'd also need one for make-based systems, for rust, for haskell, bazel, ... it's a lot :(

**Comment by L-as** [comment]
https://github.com/google/kati

**Comment by vcunat** [comment]
For reference, I suspect that the named pipes in `make` ([very new](https://git.savannah.gnu.org/cgit/make.git/commit/?id=7ad2593b2d2bb5b9332f4444d8bf93ac6f958bc6)) would simplify this a bit, but  I believe the main obstructions are elsewhere.

_ninja2nix: I think that approaches with one nix derivation per "make rule" won't scale well due to various per-derivation overheads._

**Comment by aviallon** [comment]
Might be relevant: https://github.com/olsner/jobclient

**Comment by peterwaller-arm** [comment]
I want to lend some support to this, even if maybe putting it in the stdenv for everybody is not the right place for it just yet.

I'm hitting cases where I want to spawn a large number of jobs each of which have a high degree of parallelism and a lot of memory usage within them, and a wide distribution in terms of the 'time taken to run'. Really, I want to start all of the top level jobs within a job server framework, and then let the inner jobs take tokens so that I don't hit an OOM situation. Something like this would be very useful if it were possible to opt into it. If I limit the top level parallelism, then utilization of the machine is limited. If I enlarge the machine, again, the overall utilization of the machine is not good.

Think of a top level job as ["A few long jobs" + "thousands of smaller"], each of which wanting a constant amount of RAM such that if you run multiple top level jobs (and allow them to use n-core parallelism) at once, you'll run out of RAM. If you limit to one top level job, your large machine spends the majority of its CPU time idle, since the thousands of smaller jobs complete reasonably fast leaving the few long running jobs going for a while before the next top-level job starts.

I'm finding that if I split the job up into derivation-per-piece-of-parallelism, I'm hitting limits like the fact that nix itself with the current implementation can only currently spawn something like 10 derivation builds per second, again resulting in a very poor wall clock time and CPU utilisation, and machine utilisation which does not scale with the size of the machine. 

As another point of reference, consider what happens when you end up with a very large number of small jobs. I'm having something on the order of 100k derivations. At 10/s (approx current nix limitation), with 1s-cpu-time each this means 10,000s (>2.5h) minimum wall clock to execute the jobs regardless of parallelism, where it could execute in 10 minutes if running under make on a 192 core machine. This is before you consider overheads like consulting the caches (and potentially generating a lot of traffic to cache.nixos.org if one is not careful). If I can instead split my derivations up into 10 jobs each of which running 10k processes, but allow this all to run in parallel without running out of RAM, that would be a good place to be.

**Comment by RaitoBezarius** [comment]
@ofborg test

**Comment by RaitoBezarius** [comment]
@ofborg eval

File: pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+};
+static const std::string acl_name = "system.posix_acl_access";
+static std::optional<std::vector<char>> acl;
+
+static unsigned tokens = 0;
+static std::unordered_set<Context*> poll_waiters;
+static std::unordered_map<fuse_req_t, Context*> read_waiters;
+
+static void wake_waiters()
+{
```
Comment (lf-) [inline pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp:93]: contemplating how this works under contention:

1. clients doing blocking read() will always go first
2. this looks very liable to cause spurious wakeups on poll waiters since it wakes every poller so they will thundering-herd into read().

I am assuming both of these are by design, but I do wonder if poll/async io based users would get more or less totally deprioritized by accident over blocking ones: I am assuming that if you have a loop that looks like:

```
for (;;) {
    set = poll()
    if (set[fd] is readable) {
        toks = read(fd); 
        if (toks != -1) {
             // do whatever
             break;
        } else {
             if (errno == EAGAIN) {
                 continue;
             }
        }
    }
}
```

then if you have another process that wins the race from `poll()` to `read()`, some nonblocking client would not get tokens in a pretty unfairly distributed manner.

Is this a correct assessment?

To be fair, I don't think this could be written in a way that doesn't have this problem if it is FUSE. The only way I could conceive it working is if we used unix sockets that have *actual requests* for slots and somewhat of client identity (and then can also do requesting linker slots), then we could *deterministically* wake up single clients and give them the slot they asked for.

**Review by lf- — COMMENTED** [review]
I have done a little bit of picking through this, and although I am not particularly convinced the impurity is the level of a problem suggested by others, I am not sure that the FUSE based approach supports future use cases, or has very fair semantics (unless I misunderstand).

I am also not sure what the comment means re: unix sockets and buffering and not being able to detect poll(): FUSE is in a sense turning the filesystem into a socket, with *some* kernel advantages, but it seems to me like it's plausibly about the same number of round trips to have a Unix socket ~~datagram~~stream mode protocol that looks like:

-> request token of type TYPE
<- grant token of type TYPE
-> give back token of type TYPE

in arbitrary order. Then if the socket gets closed you assume the client died. This seems basically equivalent to the FUSE implementation with fewer crimes, *assuming you control the clients*.

That said, I *do* see the need to have cursed FUSE things *somewhere* to be able to implement bridging into the make clients and have it actually work properly with some more sensible protocol, and it is probably better for the cursed FUSE things to *not* be inside the builder :(

I guess it's really just, it is kind of sad, because the actual protocol is the problem causing us suffering here, but it's in software we don't control unless we like, patch make and ninja.

**Comment by lf-** [comment]
Some interesting links I have found while looking into this:
- Failed attempt by ocaml to make a jobserver implementation: https://github.com/ocaml/dune/pull/4331 https://github.com/ocaml/opam/wiki/Spec-for-GNU-make-jobserver-support#extended-jobserver
- Kernel devs break the make jobserver with pipe changes https://lwn.net/Articles/864947/
- Ninja not merging a jobserver
  - Ninja PR that is, [according to Fuchsia devs, Busted Architecturally](https://github.com/ninja-build/ninja/issues/1139#issuecomment-1969156291), here: https://github.com/ninja-build/ninja/pull/1140
  - Fuchsia jobserver work in a fork, but I think it is not done? I can go poke someone and find that out: https://fuchsia-review.googlesource.com/q/jobserver-support

The way that Make does jobserver things seems *really* fragile in its use of syscall semantics, which is somewhat scary as the prospective implementers of such semantics. I would almost be interested in fixing the client software to use a less fragile and more featureful protocol.

**Comment by RaitoBezarius** [comment]
> The way that Make does jobserver things seems _really_ fragile in its use of syscall semantics, which is somewhat scary as the prospective implementers of such semantics. I would almost be interested in fixing the client software to use a less fragile and more featureful protocol.

Quick remark though, in the absence of a better protocol that would require adoption across build systems to be useful for the build farm, it seems like to me that having this is a stopgap solution for the current waste at the nixpkgs scale.

There are two type of outcomes:

- excessive instability stemming from the fragile jobserver concept of GNU Make: requires a removal PR
- acceptable operations for that thing: requires an upgrade PR in the future once all build systems move to a better protocol to handle this

It would be interesting to conduct more testing but it would be a shame to pass up on free benefits for the current infrastructure.

Do you think we can realistically get this somewhere first and then open up the work to a better protocol? I think there's also interesting Lix relationships notably with the whole scheduling awareness thingie and learning (can this build benefit from multiple cores, etc, etc.)

File: pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        } else {
+            std::cerr << "invalid argument for -u" << std::endl;
+            return 1;
+        }
+    }
+
+    if (!group.empty()) {
+        if (auto gr = getgrnam(group.c_str()); gr) {
+            jobserver_st.st_gid = gr->gr_gid;
+        } else if (gid_t gid; sscanf(group.c_str(), "%u%c", &gid, (char*) &gid) == 1) {
```
[the remark this answers is not in the document]
Comment (lf-) [reply pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp:426]: it feels mildly evil to do that rather than make a second scratch variable that doesn't do anything, since one could have "123a", probably? I'm not sure if it's an aliasing rule violation but it feels somewhat ehh.

File: pkgs/stdenv/generic/jscall.c [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        fprintf(stderr, "def-args command not terminated\n\n%s", usage);
+        return 1;
+    }
+    shift_once(def_args + def_args_length);
+
+    char** args = def_args + def_args_length;
+
+    struct stat st;
+    if (stat(argv[1], &st)) {
+        perror("stat() jobserver pipe");
```
[the remark this answers is not in the document]
Comment (lf-) [reply pkgs/stdenv/generic/jscall.c:135]: technically this could race if the socket vanishes in the meantime; I'm not awake enough at the moment to be certain of whether the umount would cause that symptom.

File: pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp (hunk 3 above)
Comment (lf-) [reply pkgs/by-name/ni/nixos-jobserver/nixos-jobserver.cpp:93]: could probably use a comment that says this *does* cause a thundering herd and inadvertently prioritize blocking io due to api limitation. feels like the kind of operational problem *someone* might end up running into one day and should possibly be documented as known.

**Comment by lf-** [comment]
> Quick remark though, in the absence of a better protocol that would require adoption across build systems to be useful for the build farm, it seems like to me that having this is a stopgap solution for the current waste at the nixpkgs scale.

I generally concur. I think this should be possible to delete in the future in its current design without worrying about older nixpkgs versions (that is, it does *not* introduce interfaces we promise to exist) and I don't think that this design is too problematic to ship. Already, shipping it would fix operational problems of make build systems, and I think the ways it could fail would at least be somewhat obvious (if something breaks, you cause a deadlock). 

I think someone, probably you, mentioned somewhere having Prometheus monitoring of the jobserver. I would like this but it should probably be in a separate PR for review reasons, and that would be relatively harmless since it's not in stdenv itself. Adding this would reduce my concerns about potential operational problems as we would be able to find if a worker is leaking tokens, even if that doesn't feel likely to be very possible. Or also if we could observe a 10000 long buffer of who things got assigned to in relation to time (and generate a statemap, if i have my terms not mixed up) that we could pull via metrics api to see how much the unfairness is a problem in practice.

**Comment by RaitoBezarius** [comment]
> > Quick remark though, in the absence of a better protocol that would require adoption across build systems to be useful for the build farm, it seems like to me that having this is a stopgap solution for the current waste at the nixpkgs scale.
> 
> I generally concur. I think this should be possible to delete in the future in its current design without worrying about older nixpkgs versions (that is, it does _not_ introduce interfaces we promise to exist) and I don't think that this design is too problematic to ship. Already, shipping it would fix operational problems of make build systems, and I think the ways it could fail would at least be somewhat obvious (if something breaks, you cause a deadlock).
> 
> I think someone, probably you, mentioned somewhere having Prometheus monitoring of the jobserver. I would like this but it should probably be in a separate PR for review reasons, and that would be relatively harmless since it's not in stdenv itself. Adding this would reduce my concerns about potential operational problems as we would be able to find if a worker is leaking tokens, even if that doesn't feel likely to be very possible. Or also if we could observe a 10000 long buffer of who things got assigned to in relation to time (and generate a statemap, if i have my terms not mixed up) that we could pull via metrics api to see how much the unfairness is a problem in practice.

Awesome, thank you for the thorough review <3. Will do that.

**Comment by RaitoBezarius** [comment]
Addressed most of your comments @lf-, please take another look.

File: doc/stdenv/stdenv.chapter.md [hunk 6]
```diff
@@ -1355,6 +1355,30 @@ name="/nix/store/9s9r019176g7cvn2nvcw41gsp862y6b4-coreutils-8.24"
...
+ - If a global jobserver is available, runs `<command> <args>` with a GNU Make compatible jobserver made
+   available in `MAKEFLAGS`. The client is automatically limited to `NIX_BUILD_CORES` tokens.
+ - If no global jobserver is available, runs `<command> <defArgs> <args>` instead.
+
+\<command\>, \<defArgs\> and \<args\> may be more than one shell word each, hence they are separated by
+`----`. \<command\> and \<defArgs\> may not contain `----`. Each set except \<command\> may be empty,
+but separators must always be given.
+
+**Note**:
+The global jobserver must be bound to `/jobserver` in the Nix sandbox to be found by this function.
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply doc/stdenv/stdenv.chapter.md:1371]: Last commit performs the move under `/build-support/jobserver`

File: pkgs/stdenv/linux/default.nix [hunk 7]
```diff
@@ -631,6 +631,18 @@ in
       extraAttrs = {
         inherit bootstrapTools;
         shellPackage = prevStage.bash;
+
+        jscall = derivation {
```
Comment (MinerSebas) [inline pkgs/stdenv/linux/default.nix:635]: Does this need to be abbreviated?
Initially i read the js as javascript, which made no sense.
Only in the context of this PR would i recognize this as standing for jobserver.
For this reason i would recommend against the js abbreviation.

Comment (RaitoBezarius) [reply pkgs/stdenv/linux/default.nix:635]: I think it's an internal piece of code, most users will never encounter it, but I can rename it if you want.

Comment (MinerSebas) [reply pkgs/stdenv/linux/default.nix:635]: A valid argument.
As I don't work on the stdenv, I say treat this as a nitpick, unless someone else more involved also advocates for changing it.

Comment (RaitoBezarius) [reply pkgs/stdenv/linux/default.nix:635]: I added a comment, but I would rather avoid modifying the source code or modifying the attrnames as this will cause a mass rebuild and makes testing annoying for me.

We can rename it in a next staging cycle if this is a real problem.

File: nixos/modules/services/misc/nixos-jobserver.nix [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        touch ${tokenFile}
+      '';
+
+      script = ''
+        nixos-jobserver \
+          -t ${toString cfg.tokens} \
+          -u ${escapeShellArg cfg.owner} \
+          -g ${escapeShellArg cfg.group} \
+          ${tokenFile}
+      '';
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply nixos/modules/services/misc/nixos-jobserver.nix:63]: Made it `exec`

**Comment by RaitoBezarius** [comment]
Expectations: I intend to go for a merge once I re-confirm that I can build reproducibly well the NixOS minimal ISO and that the jobserver *is used*. I deployed it on one of my 128T remote builder so far.

**Comment by aviallon** [comment]
I really can't wait for this to land. Thanks @pennae and @RaitoBezarius
